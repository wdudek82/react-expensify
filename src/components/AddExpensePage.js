import React from 'react';

const AddExpensePage = () => (
  <div>
    This is Add Expense Page
  </div>
);

export default AddExpensePage;
