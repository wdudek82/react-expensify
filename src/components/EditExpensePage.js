import React from 'react';

const EditExpensePage = (props) => {
  console.log(props);
  const id = props.match.params.id;

  return (
    <div>
      <h2>Editing id: {id}</h2>
      Edit Expense Page
    </div>
  );
};

export default EditExpensePage;