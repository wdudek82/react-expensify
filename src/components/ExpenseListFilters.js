import React from 'react';
import { connect } from 'react-redux';
import moment from 'moment';
import { DateRangePicker } from 'react-dates';
import {
  setTextFilter,
  setSortByAmountFilter,
  setSortByDateFilter,
  setStartDateFilter,
  setEndDateFilter
} from '../actions/filters';


export class ExpenseListFilters extends React.Component {
  state = {
    calendarFocused: null
  }

  onTextChange = (e) => {
    this.props.setTextFilter(e.target.value);
  };
  
  onSortChange = (e) => {
    const selected = e.target.value;
    if (selected === 'date') {
      this.props.setSortByDateFilter()
    } else if (selected === 'amount') {
      this.props.setSortByAmountFilter();
    }
  }

  onDatesChange = ({ startDate, endDate }) => {
    this.props.setStartDateFilter(startDate);
    this.props.setEndDateFilter(endDate);
  }

  onFocusChange = (calendarFocused) => {
    this.setState((prevState) => ({ calendarFocused }));
  }

  render() {
    return (
      <div>
        <input
          type="text"
          value={this.props.filters.text}
          onChange={(e) => this.onTextChange(e)}
        />
        <select
          value={this.props.filters.sortBy}
          onChange={(e) => this.onSortChange(e)}>
          <option value="date">date</option>
          <option value="amount">amount</option>
        </select>
        <DateRangePicker
          startDate={this.props.filters.startDate} // momentPropTypes.momentObj or null,
          endDate={this.props.filters.endDate} // momentPropTypes.momentObj or null,
          onDatesChange={this.onDatesChange} // PropTypes.func.isRequired,
          focusedInput={this.state.calendarFocused} // PropTypes.oneOf([START_DATE, END_DATE]) or null,
          onFocusChange={this.onFocusChange} // PropTypes.func.isRequired,
          numberOfMonths={1}
          displayFormat={'DD/MM/YYYY'}
          isOutsideRange={() => false}
          showClearDates={true}
        />
      </div>
    )
  }
}

const mapStateToProps = (state) => ({
  filters: state.filters
});

const mapDispatchToProps = (dispatch, props) => ({
  setTextFilter: (text) => {
    dispatch(
      setTextFilter(text)
    )
  },
  setSortByAmountFilter: () => {
    dispatch(
      setSortByAmountFilter()
    )
  },
  setSortByDateFilter: () => {
    dispatch(
      setSortByDateFilter()
    )
  },
  setStartDateFilter: (startDate) => {
    dispatch(
      setStartDateFilter(startDate)
    )},
  setEndDateFilter: (endDate) => {
    dispatch(
      setEndDateFilter(endDate)
    )
  },
});

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);