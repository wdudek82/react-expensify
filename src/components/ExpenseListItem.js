import React from 'react';
import { Link } from 'react-router-dom';
import moment from 'moment';

const ExpenseListItem = ({ id, description, amount, createdAt }) => (
  <div key={id}>
    <Link to={`/edit/${id}`}>
      <h3>{description}</h3>
    </Link>
    <p>{amount} - created: {createdAt}</p>
  </div>
);

export default ExpenseListItem;