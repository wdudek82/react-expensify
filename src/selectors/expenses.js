const getVisibleExpenses = (
  expenses,
  { text, sortBy, startDate, endDate }
) => {
  return expenses.filter(expense => {
    const startDateMatch = typeof startDate !== 'number' || expense.createdAt >= startDate;
    const endDateMatch = typeof endDate !== 'number' || expense.createdAt <= endDate;
    
    const expenseDescription = expense.description.toLowerCase();
    const includesText = expenseDescription.includes(text.toLowerCase());

    return startDateMatch && endDateMatch && includesText;
  }).sort((a, b) => {
    if (sortBy === 'date') {
      return a.createdAt < b.createdAt;
    } else if (sortBy === 'amount') {
      return a.amount < b.amount;
    }
  });
};

export default getVisibleExpenses;