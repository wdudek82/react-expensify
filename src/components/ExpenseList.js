import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';

const ExpenseList = (props) => {
  const expenses = props.expenses.map((expense, index) => (
    <ExpenseListItem
      key={`expense-${index}`}
      optionKey={`expense-${index}`}
      expense={expense}
    />
  ));

  return (
    <div>
      <h1>Expense List</h1>
      {expenses}
    </div>
  );
};

const mapStateToProps = (state) => {
  return {
    expenses: state.expenses,
    filters: state.filters
  };
};

export default connect(mapStateToProps)(ExpenseList);

// const ConnectedExpenseList = connect((state) => {
//   return {
//     expenses: state.expenses
//   };
// })(ExpenseList);
// 
// export default ConnectedExpenseList;