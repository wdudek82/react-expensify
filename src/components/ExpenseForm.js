import React from 'react';
import { connect } from 'react-redux';
import { addExpense } from '../actions/expenses';


class ExpenseForm extends React.Component {
  state = {
    description: '',
    amount: 0,
    createdAt: 0,
    note: ''
  }

  handleSubmitForm = (e) => {
    e.preventDefault();
    this.props.dispatch(
      addExpense({ ...this.state })
    );
    this.setState(() => ({
      description: '',
      amount: 0,
      createdAt: 0,
      note: ''
    }));
  };

  onDescriptionChange = (e) => {
    const description = e.target.value;
    this.setState(() => ({ description }));
  };

  onAmountChange = (e) => {
    const amount = parseInt(e.target.value) || 0;
    this.setState(() => ({ amount }));
  }

  onDateChange = (e) => {
    const createdAt = parseInt(e.target.value) || 0;
    this.setState(() => ({ createdAt }));
  }

  onNoteChange = (e) => {
    const note = e.target.value;
    this.setState(() => ({ note }));
  };

  render() {
    return (
      <div>
        <h1>Add Expense</h1>
        <form onSubmit={this.handleSubmitForm}>
          <div>
            <input
              type="text"
              placeholder="Description"
              autoFocus
              value={this.state.description}
              onChange={this.onDescriptionChange}
            />
          </div>
          <div>
            <input
              type="number"
              placeholder="Amount"
              value={this.state.amount}
              onChange={this.onAmountChange}
            />
          </div>
          <div>
            <input
              type="number"
              placeholder="Date"
              value={this.state.createdAt}
              onChange={this.onDateChange}
            />
          </div>
          <div>
            <textarea
              placeholder="Add a note for your expense (optional)"
              value={this.state.note}
              onChange={this.onNoteChange}
            ></textarea>
          </div>
          <input type="submit" value="Add Expense" />
        </form>
      </div>
    );
  }
}

export default connect()(ExpenseForm);
