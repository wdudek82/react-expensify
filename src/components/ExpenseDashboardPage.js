import React from 'react';
import ExpenseList from './ExpenseList';
import ExpenseListFilters from './ExpenseListFilters';

class ExpenseDashboardPage extends React.Component {
  render() {
    return (
      <div>
        <ExpenseListFilters />
        <ExpenseList />
      </div>
    );
  };
}

export default ExpenseDashboardPage;