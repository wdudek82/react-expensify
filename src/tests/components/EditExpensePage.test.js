import React from 'react';
import { shallow } from 'enzyme';
import { EditExpensePage } from '../../components/EditExpensePage';
import expenses from '../fixtures/expenses';

let expense,
    editExpense,
    removeExpense,
    history,
    wrapper;

beforeEach(() => {
  expense = expenses[1];
  editExpense = jest.fn();
  removeExpense = jest.fn();
  history = { push: jest.fn() };
  wrapper = shallow(
    <EditExpensePage
      editExpense={editExpense}
      removeExpense={removeExpense}
      expense={expense}
      history={history}
    />
  );
});

it('should render correctly EditExpensePage component', () => {
  expect(wrapper).toMatchSnapshot();
});

it('should handle onSubmit', () => {
  wrapper.find('ExpenseForm').prop('onSubmit')(expense);
  expect(history.push).toHaveBeenLastCalledWith('/');
  expect(editExpense).toHaveBeenLastCalledWith(expense.id, expense);
});

if('should handle handleRemoveExpense', () => {
  wrapper.find('button').simulate('click')();
  expect(history.push).toHaveBeenLastCalledWith('/');
  expect(removeExpense).toHaveBeenLastCalledWith({
    id: expense.id
  });
});