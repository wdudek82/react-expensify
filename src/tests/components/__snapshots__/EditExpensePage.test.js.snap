// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly EditExpensePage component 1`] = `
<div>
  <h1>
    Edit Expense
  </h1>
  <ExpenseForm
    expense={
      Object {
        "amount": 100.95,
        "createdAt": 172800000,
        "description": "Water Bill",
        "id": "2",
        "note": "First note",
      }
    }
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
  >
    Remove
  </button>
</div>
`;
